src/namespace.rs

```diff
@@ -0,0 +1,167 @@
+use std::convert::Into;
+use std::env::current_dir;
+use std::env::set_current_dir;
+use std::fs;
+use std::fs::File;
+use std::io;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::os::fd::AsFd;
+use std::os::fd::AsRawFd;
+use std::os::unix::fs::MetadataExt;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use libc::setns;
+use libc::CLONE_NEWNS;
+
+use crate::Error;
+use crate::ErrorExt;
+use crate::Pid;
+
+pub(crate) struct NsInfo {
+    tgid: Pid,
+    nstgid: Pid,
+    need_setns: bool,
+    mntns_path: Option<PathBuf>,
+    oldns: File,
+    // From https://github.com/torvalds/linux/commit/b01c1f69c8660eaeab7d365cd570103c5c073a02, we see
+    // once finished we setns to old namespace, which also sets the current working directory (cwd) to "/",
+    // trashing the cwd we had. So adding the current working directory to be part of `NsInfo` and restoring
+    // it in the `Drop` call.
+    oldcwd: PathBuf,
+}
+
+fn get_nspid(pid: Pid) -> Result<(Pid, Pid), Error> {
+    let fname = format!("/proc/{pid}/status");
+    let file = File::open(&fname).context("faild to open `{fname}`")?;
```

Comment:
`context` won't do any string interpolation (formatting), as it just accepts a string directly. You'll have to `format` explicitly:
```suggestion
    let file = File::open(&fname).with_context(|| format!("faild to open `{fname}`"))?;
```